src/CheckStrings.jl

```diff
@@ -0,0 +1,200 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+## Functions to check validity of UTF-8, UTF-16, and UTF-32 encoded strings,
+#  and also to return information necessary to convert to other encodings
+
+module CheckStrings
+
+using Encodings
+
+export check_string
+export is_surrogate_lead, is_surrogate_trail, is_surrogate_codeunit, is_valid_continuation
+
+using Base.UTF_ERR_SHORT, Base.UTF_ERR_CONT,Base.UTF_ERR_LONG,
+      Base.UTF_ERR_NOT_LEAD, Base.UTF_ERR_NOT_TRAIL,
+      Base.UTF_ERR_SURROGATE, Base.UTF_ERR_MISSING_SURROGATE, Base.UTF_ERR_INVALID
+
+is_surrogate_lead(c::Unsigned) = ((c & ~0x003ff) == 0xd800)
+is_surrogate_trail(c::Unsigned) = ((c & ~0x003ff) == 0xdc00)
+is_surrogate_codeunit(c::Unsigned) = ((c & ~0x007ff) == 0xd800)
+is_valid_continuation(c) = ((c & 0xc0) == 0x80)
+
+## Options for check_string_* functions
+
+const UTF_NO_LONG_NULL = 1      ##< don't accept 0xc0 0x80 for '\0'
+const UTF_NO_SURROGATES = 2     ##< don't accept surrogate pairs in UTF-8/UTF-32
+const UTF_ACCEPT_LONG = 4       ##< accept long encodings (other than long null in UTF-8)
+
+const UTF_LONG = 1              ##< Long encodings are present
+const UTF_LATIN1 = 2            ##< characters in range 0x80-0xFF present
+const UTF_UNICODE2 = 4          ##< characters in range 0x100-0x7ff present
+const UTF_UNICODE3 = 8          ##< characters in range 0x800-0xd7ff, 0xe000-0xffff
+const UTF_UNICODE4 = 16         ##< non-BMP characters present
+const UTF_SURROGATE = 32        ##< surrogate pairs present
+
+## Get a UTF-8 continuation byte, give error if invalid, return updated character value
+@inline function get_continuation(ch::UInt32, byt::UInt8, pos)
+    !is_valid_continuation(byt) && throw(UnicodeError(UTF_ERR_CONT, pos, byt))
+    (ch << 6) | (byt & 0x3f)
+end
+
+CodeUnitType = Union(UInt8, UInt16, UInt32, Char)
+CodeUnitC = Union(Vector{Char}, AbstractVector{Char}, AbstractArray{Char})
+CodeUnit8 = Union(Vector{UInt8}, AbstractVector{UInt8}, AbstractArray{UInt8})
```

Comment:
No, I want to deal with the code units, not the logical characters.
The interface of `AbstractString`s like `UTF8String` and `UTF16String` is to have `next()` check the validity (which won't work, because we need to allow some things that aren't valid, later on, to use a replacement string, or to handle "semi-valid" things like the `Modified UTF-8` or `CESU-8` encodings.